### ISSUE TITLE
Single-Valued fields can't be used in conditionals in description_templates.

### ISSUE THREAD

**Issue by dssinger** [RhetTbull/osxphotos#506]
[With](![tempImagezeHmPd](https://user-images.githubusercontent.com/1817903/130376723-42e6a20e-b618-439b-bbe1-0599efe8358f.gif)) a description_template of:
````
"{title,No Title} and {descr,No Descr}"
````
and the attached photo (with no title and no description), the result is " and " - the title and description fields are being treated as if they were present even though they aren't.

It looks like the problem was introduced with changes #485 and #486 when you added
````
 options = dataclasses.replace(
            self._render_options, expand_inplace=True, inplace_sep=", "
        )
````
just before the call to `render_template`.  

I've tried to fix the problem in `render_template_string` by ignoring the separator and expand_inplace for single-valued fields. 
````
            if not singlevalueneeded and (self.expand_inplace or delim is not None):
                sep = delim if delim is not None else self.inplace_sep
                vals = [sep.join(sorted(vals))]
````

I'll submit a pull request.

Thanks.

**Comment by dssinger** [comment]
![tempImageR214Te](https://user-images.githubusercontent.com/1817903/130376776-e655037f-98bd-4748-ad16-145419a0daf0.gif)

**Comment by dssinger** [comment]
My fix isn't quite right; it appends an underscore to the "false" value of a boolean, like so:

````
description-template "{title}{title?{descr?{descr != OLYMPUS DIGITAL CAMERA? - ,oly1},nodesc},notitle}{descr != OLYMPUS DIGITAL CAMERA?{descr},oly }" 
````
yields `_notitle` if there's no title or descriptor or `{title}nodesc_` if there's no descriptor. 

I'm not quite sure where to go with this now - Rhet, I hope this gives you enough info to fix the problem properly.

**Comment by RhetTbull** [comment]
Thanks @dssinger for attempting a PR. I'm traveling but will take a look when I get home. This is a fundamental feature of the template system so I'm surprised I didn't have a test to catch this--will fix that too!

**Comment by RhetTbull** [comment]
@dssinger I think this is fixed in [v0.42.74](https://github.com/RhetTbull/osxphotos/releases/tag/v0.42.74).  Let me know if you're still having problems.  Glad you caught this -- you pointing me to the commit for #485, #486 really helped narrow down the issue.

**Comment by dssinger** [comment]
That fixed the {title,No title} issue, but an underscore is being appended to the "false" value for a conditional test, like so:

````
 --description-template "{title}{title?{descr?{descr != OLYMPUS DIGITAL CAMERA? - ,},},}{descr != OLYMPUS DIGITAL CAMERA?{descr},}" 
````

yields `__` for the test photo above (and similarly adds an underscore to anything specified for the false side of a conditional).

**Comment by RhetTbull** [comment]
> but an underscore is being appended to the "false" value for a conditional test,

I did a test for this and was not able to replicate it

>"{title}{title?{descr?{descr != OLYMPUS DIGITAL CAMERA? - ,},},}{descr != OLYMPUS DIGITAL CAMERA?{descr},}" 

For your image with no title and no description, this template evaluates to `__` (double underscore). See breakdown below:

There are three separate template statements here:

`{title}` == "_"
title is null and no default specified, so the value will be _

`{title?{descr?{descr != OLYMPUS DIGITAL CAMERA? - ,},},}` == ""
title is null so so `title?` evaluates to false so the false value will be used which in this case is null string (nothing after the trailing comma)

`{descr != OLYMPUS DIGITAL CAMERA?{descr},}` == "_"

descr is null and so `descr  != OLYMPUS DIGITAL CAMERA?` will be True and thus the true value of the conditional will be used which in this case is `{descr}`.  But since description is null and no default value was provided, the `_` default will be used.

Thus the expected value (assuming title and descr are null) for your sample template string is `"_"` + `""` + `"_"` = `"__"`

**Comment by dssinger** [comment]
What I'm trying to do is set the exported photo's description to "{title} - {descr}" if both are set in the Photos library, to "{title}" if there's only a title, to "{descr}" if there's only a description, or to nothing if neither is set (ignore the "OLYMPUS DIGITAL CAMERA" part for now).  This is the same as the example in the README showing the finder comment being set that way (except with a " - " instead of a newline) - and it works for the finder comment, but not for the description.

I'm attaching a small test photo library (zipped) with four photos with all combinations of title and description and a script that uses osxphotos to set the description and the finder comment for the exported photos using the same template, as well as a zip file of the exported photos.  

Photo library:
[testlib.zip](https://github.com/RhetTbull/osxphotos/files/7042710/testlib.zip)

Test script (zipped to make GitHub happy):
[testit.zip](https://github.com/RhetTbull/osxphotos/files/7042716/testit.zip)

Exported photos:
[results.zip](https://github.com/RhetTbull/osxphotos/files/7042709/results.zip)

I really appreciate your help with this!

**Comment by RhetTbull** [comment]
I'm traveling and don't have access to my Mac but will take a look this weekend. Thanks for the sample library -- that'll help!

**Comment by RhetTbull** [comment]
I have this partially figured out.  The template in the Tutorial is actually wrong.  The template should be:
`"{title,}{title?{descr? - ,},}{descr,}"`

instead of 

`"{title}{title?{descr? - ,},}{descr}"`

Note the `,` after `{title` and `{descr` which means "if value is blank, use an empty string for the value instead of the default _ "; without the commas, if title or descr is blank the standard default value "_" is used which causes the values you saw for --description-template.

This doesn't happen for Finder comment because the code does some special handling of the standard default value which results in an empty string being used as the default value.  So the bug is actually in this part of the code.  I need to go figure out why I did that in the first place then fix it.  There's another bug that results in the white space around ` - ` also being stripped for the Finder comments but not the description (see below).

Results when used with the corrected template:

```
[I] ➜ exiftool -G -Title -Description ~/Desktop/Test/2021-08-24/*.jpg
======== /Users/rhet/Desktop/Test/2021-08-24/P1050198.jpg
[XMP]           Title                           : Title
[XMP]           Description                     : Title - Description
======== /Users/rhet/Desktop/Test/2021-08-24/P1050200.jpg
[XMP]           Title                           : Title
[XMP]           Description                     : Title
======== /Users/rhet/Desktop/Test/2021-08-24/P1050202.jpg
[MakerNotes]    Title                           :
[XMP]           Description                     : Description
======== /Users/rhet/Desktop/Test/2021-08-24/P1050204.jpg
[MakerNotes]    Title                           :
    4 image files read

[I] ➜ osxmetadata --list ~/Desktop/Test/2021-08-24/*.jpg
/Users/rhet/Desktop/Test/2021-08-24/P1050198.jpg:
findercomment               com.apple.metadata:kMDItemFinderComment            = Title-Description
/Users/rhet/Desktop/Test/2021-08-24/P1050200.jpg:
findercomment               com.apple.metadata:kMDItemFinderComment            = Title
/Users/rhet/Desktop/Test/2021-08-24/P1050202.jpg:
findercomment               com.apple.metadata:kMDItemFinderComment            = Description
/Users/rhet/Desktop/Test/2021-08-24/P1050204.jpg:
```

**Comment by RhetTbull** [comment]
I added some fixes for "--strip" to [v0.42.77](https://github.com/RhetTbull/osxphotos/releases/tag/v0.42.77) and I think these, with the combination of correcting the template to `"{title,}{title?{descr? - ,},}{descr,}"` should fix your problem. Running your test script produces what I think is your expected output.  Please let me know if this does/does not fix your issue.

```
exiftool -G -Title -Description ~/Desktop/Test/2021-08-24/*.jpg
======== /Users/rhet/Desktop/Test/2021-08-24/P1050198.jpg
[XMP]           Title                           : Title
[XMP]           Description                     : Title - Description
======== /Users/rhet/Desktop/Test/2021-08-24/P1050200.jpg
[XMP]           Title                           : Title
[XMP]           Description                     : Title
======== /Users/rhet/Desktop/Test/2021-08-24/P1050202.jpg
[MakerNotes]    Title                           :
[XMP]           Description                     : Description
======== /Users/rhet/Desktop/Test/2021-08-24/P1050204.jpg
[MakerNotes]    Title                           :
    4 image files read

[I] ➜ osxmetadata --list ~/Desktop/Test/2021-08-24/*.jpg
/Users/rhet/Desktop/Test/2021-08-24/P1050198.jpg:
findercomment               com.apple.metadata:kMDItemFinderComment            = Title - Description
/Users/rhet/Desktop/Test/2021-08-24/P1050200.jpg:
findercomment               com.apple.metadata:kMDItemFinderComment            = Title
/Users/rhet/Desktop/Test/2021-08-24/P1050202.jpg:
findercomment               com.apple.metadata:kMDItemFinderComment            = Description
/Users/rhet/Desktop/Test/2021-08-24/P1050204.jpg:
```

**Comment by dssinger** [comment]
Yes, that fixes the problem - even with the more complicated template I need to get rid of OLYMPUS DIGITAL CAMERA:
````
export TEMPLATE="{title,}{title?{descr?{descr != OLYMPUS DIGITAL CAMERA? - ,},},}{descr?{descr != OLYMPUS DIGITAL CAMERA?{descr},},}"
````

I found two other oddities while verifying the fix (neither of which will cause me any trouble) - I will close this problem and open new ones for your consideration.

Thanks very much!
